pkg/corerp/api/v20231001preview/environment_conversion.go

```diff
@@ -408,35 +420,118 @@ func fromRecipePropertiesClassificationDatamodel(e datamodel.EnvironmentRecipePr
 	return nil
 }
 
-func toRecipeConfigTerraformProvidersDatamodel(config *RecipeConfigProperties) map[string][]datamodel.ProviderConfigProperties {
+func toRecipeConfigTerraformProvidersDatamodel(config *RecipeConfigProperties) (map[string][]datamodel.ProviderConfigProperties, error) {
 	if config.Terraform == nil || config.Terraform.Providers == nil {
-		return nil
+		return nil, nil
 	}
 
 	dm := map[string][]datamodel.ProviderConfigProperties{}
 	for k, v := range config.Terraform.Providers {
 		dm[k] = []datamodel.ProviderConfigProperties{}
 
 		for _, providerAdditionalProperties := range v {
+			var addnlProperties map[string]any
+			var convertedSecrets map[string]datamodel.SecretReference
+
+			convertedSecrets = getSecretsFromProviderProperties(providerAdditionalProperties.Secrets)
+			addnlProperties = providerAdditionalProperties.AdditionalProperties
+
+			// Account for possibility of nested additional properties during unmarshalling
+			// https://json-schema.org/understanding-json-schema/reference/object#additionalproperties
+			if tmpAddnlProperties, ok := addnlProperties["additionalProperties"]; ok && addnlProperties["additionalProperties"] != nil {
+				if tmpMap, ok := tmpAddnlProperties.(map[string]any); ok {
+					addnlProperties = tmpMap
+				} else {
+					return nil, v1.ErrInvalidModelConversion
+				}
+			}
+
+			// check if additional properties contain Secrets
+			// get AdditionalProperties from providerAdditionalProperties excluding the 'secrets' key.
+			for key, value := range addnlProperties {
+				if key == "secrets" {
+					if tmpSecret, ok := value.(map[string]*SecretReference); !ok {
+						return nil, v1.ErrInvalidModelConversion
+					} else {
+						propSecrets := getSecretsFromProviderProperties(tmpSecret)
+
+						// Add the secrets to the convertedSecrets map
+						for secretKey, secretValue := range propSecrets {
+							if convertedSecrets == nil {
+								convertedSecrets = map[string]datamodel.SecretReference{}
+							}
```

Comment:
This code is getting difficult to follow with abbreviations being used which I'm sure are obvious to you while writing the code but I don't know how to interpret `tmpSecret` and `propSecret`. Can you look into making it more readable?

Reply:
This has been removed. pls take another look.